docs/source/examples/managed-jobs.rst

```diff
@@ -179,17 +302,74 @@ Below is an example of mounting a bucket to :code:`/checkpoint`.
+
+For other types of workloads, you can implement a similar mechanism as long as you can store the program state to/from disk.
+
+
+.. _failure-recovery:
+
+Jobs restarts on user code failure
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Preemptions or hardware failures will be auto-recovered, but **by default, user code failures (non-zero exit codes) are not auto-recovered**.
+
+In some cases, you may want a job to automatically restart even if it fails in application code. For instance, if a training job crashes due to an NVIDIA driver issue or NCCL timeout, it should be recovered. To specify this, you
+can set :code:`max_restarts_on_errors` in :code:`resources.job_recovery` in the job YAML file.
+
+.. code-block:: yaml
+
+  resources:
+    accelerators: A100:8
+    job_recovery:
+      # Restart the job up to 3 times on user code errors.
+      max_restarts_on_errors: 3
+
+This will restart the job, up to 3 times (for a total of 4 attempts), if your code has any non-zero exit code. Each restart runs on a newly provisioned temporary cluster.
+
+Here's how various kinds of failures will be handled by SkyPilot:
+
+.. list-table::
+   :widths: 1 2
+   :header-rows: 0
+
+   * - **User code fails (**:code:`setup` **or** :code:`run` **commands have non-zero exit code):**
+     - If :code:`max_restarts_on_errors` is set, restart up to that many times. If :code:`max_restarts_on_errors` is not set, or we run out of restarts, set the job to :code:`FAILED` or :code:`FAILED_SETUP`.
+   * - **Instances are preempted or underlying hardware fails:**
+     - Tear down the old temporary cluster and provision a new one in another region, then restart the job.
+   * - **Can't find available resources due to cloud quota or capacity restrictions:**
+     - Try other regions and other clouds indefinitely until resources are found.
+   * - **Cloud config/auth issue or invalid job configuration:**
+     - Mark the job as :code:`FAILED_PRECHECKS` and exit. Won't be retried.
+
+To see the logs of user code (:code:`setup` or :code:`run` commands), use :code:`sky jobs logs <job_id>`. If there is a provisioning or recovery issue, you can see the provisioning logs by running :code:`sky jobs logs --controller <job_id>`.
+
+.. tip::
+  Under the hood, SkyPilot uses a "controller" to provision, monitor, and recover the underlying temporary clusters. To learn more about the controller, see :ref:`jobs-controller`.
 
 
 .. _spot-jobs-end-to-end:
 
-An End-to-End Example
-~~~~~~~~~~~~~~~~~~~~~
+Examples
+--------
+
+BERT end-to-end
+~~~~~~~~~~~~~~~
 
 Below we show an `example <https://github.com/skypilot-org/skypilot/blob/master/examples/spot/bert_qa.yaml>`_ for fine-tuning a BERT model on a question-answering task with HuggingFace.
 
+This example:
+
+- has SkyPilot find a V100 instance on any cloud,
+- uses spot instances to save cost, and
+- uses checkpointing to recover preempted jobs quickly.
+
 .. code-block:: yaml
```

Comment:
Add to beginning of L421 to make it read better "The highlighted lines add in bucket mounting (for saving and reading checkpoints) and instruct the training framework to write checkpoints there." Then make L421 a bit more concise?

periodically -> periodic

(And I know these are remnants from before!)